working/augmentation-libraries/feature-specification.md

```diff
@@ -622,311 +626,97 @@ It is a compile-time error if:
 
-When augmenting an enum value, no constructor invocation should be provided.
-The original value is always used, and the explicit constructor invocation (if
-present) should not be copied.
+A class-like augmentation can add *new* variables (i.e. instance or static
+fields) to the type being augmented:
 
-New enum values may be defined in an augmenting enum, and they will be appended
-to the current values of the declaration in augmentation application order.
-
-Augmenting an existing enum value never changes the order in which it appears in
-`values`.
+```dart
+class C {}
 
-For example:
+augment class C {
+  int x = 3;
 
+  static int y = 4;
+}
 ```
-// main.dart
-part 'a.dart';
-part 'c.dart';
 
-enum A {
-  first,
-  second.custom(1);
+Variable declarations themselves can't be directly augmented, with the exception
+of abstract fields.
 
-  final int b;
+*A variable declaration implicitly has code for the synthesized getter and
+setter that access and modify the underlying backing storage. Since we don't
+allow augmentations to replace code, that implies that augmentations can't
+change variables. So we don't allow them to be augmented.*
 
-  const A() : b = 0;
+#### Abstract fields
 
-  const A.custom(this.b);
-}
-}
+Dart supports `abstract` field declarations. They are syntactic sugar for
+declaring an abstract getter with an optional abstract setter if the variable is
+non-final. They don't actually declare a variable with any backing storage.
 
-// a.dart
-part of 'main.dart';
-part 'b.dart';
+Because abstract variables are effectively abstract getter and setter
+declarations, they can be augmented and used in augmentations just like function
+declarations:
 
-augment enum A {
-  third;
+*Examples:*
 
-  /// Some doc comment
-  augment first; // This is still `first` in values.
+```dart
+class C {
+  // Augment an abstract variable with a getter:
+  abstract final int a;
+  augment int get a => 1; // OK.
 
-  @someAnnotation
-  augment second; // Don't repeat the argument list, original is used.
-}
+  // Augment an abstract variable with a getter and setter:
+  abstract int b;
+  augment int get b => 1; // OK.
+  augment set b(int value) {} // OK.
 
-// b.dart
-part of 'a.dart';
+  // Augment a getter with an abstract variable:
+  int get c;
 
-augment enum A {
-  fourth;
+  @someMetadata
+  augment abstract final int c; // (Not very useful, but valid.)
 }
+```
 
-// c.dart
-part of 'main.dart';
-
-augment enum A {
-  fifth;
+### Augmenting enum members
 
-  // Error, enum value augmentations cannot have an explicit constructor
-  // invocation.
-  augment third.custom(3);
-}
-```
+An augmentation of an enum type can add new members to the enum, including new
+enum values. Enum values are appended in augmentation application order.
 
-Then `A.values` is `[A.first, A.second, A.third, A.fourth, A.fifth]`.
+Enum values themselves can't be augmented since they are essentially constant
+variables and variables can't be augmented.
```

Comment:
_(But the static getters can, even if only to add metadata.)_

Reply:
In this PR, the pitch is that variables can't be augmented *at all*, even by an abstract getter that only adds metadata to the variable. We could allow this, but then it needs to be restricted such that the augmenting getter or setter has to be abstract (incomplete). If you augment a variable with a setter, does the metadata get copied to the variable's implicit getter too?

We could work this out, but I suspect it starts to get special-case-heavy and I don't think it's likely to be very valuable since all you could do is add metadata. So I'm inclined to just leave it out entirely. Variables are *overridable* in the sense of polymorphism and uniform access and all that. But when it comes to augmentations, they're fixed. We could always loosen this in the future if we get a use case.